package/MDAnalysis/analysis/base.py

```diff
@@ -332,65 +401,300 @@ def _setup_frames(self, trajectory, start=None, stop=None, step=None,
 
 
         .. versionchanged:: 1.0.0
             Added .frames and .times arrays as attributes
-
+            
         .. versionchanged:: 2.2.0
             Added ability to iterate through trajectory by passing a list of
             frame indices in the `frames` keyword argument
-
+            
+        .. versionchanged:: 2.8.0
+            Split function into two: :meth:`_define_run_frames` and
+            :meth:`_prepare_sliced_trajectory`: first one defines the limits
+            for the whole run and is executed once during :meth:`run` in
+            :meth:`_setup_frames`, second one prepares sliced trajectory for
+            each of the workers and gets executed twice: one time in
+            :meth:`_setup_frames` for the whole trajectory, second time in
+            :meth:`_compute` for each of the computation groups.
         """
-        self._trajectory = trajectory
-        if frames is not None:
-            if not all(opt is None for opt in [start, stop, step]):
-                raise ValueError("start/stop/step cannot be combined with "
-                                 "frames")
-            slicer = frames
-        else:
-            start, stop, step = trajectory.check_slice_indices(start, stop,
-                                                               step)
-            slicer = slice(start, stop, step)
-        self._sliced_trajectory = trajectory[slicer]
-        self.start = start
-        self.stop = stop
-        self.step = step
-        self.n_frames = len(self._sliced_trajectory)
-        self.frames = np.zeros(self.n_frames, dtype=int)
-        self.times = np.zeros(self.n_frames)
+        slicer = self._define_run_frames(trajectory, start, stop, step, frames)
+        self._prepare_sliced_trajectory(slicer)
 
     def _single_frame(self):
         """Calculate data from a single frame of trajectory
 
         Don't worry about normalising, just deal with a single frame.
+        Attributes accessible during your calculations:
+
+          - ``self._frame_index``: index of the frame in results array
+          - ``self._ts`` -- frame timestamp value
+          - ``self._sliced_trajectory`` -- trajectory that you're iterating over
+          - ``self.results`` -- :class:`MDAnalysis.analysis.results.Results` instance
+            holding run results initialized in :meth:`_prepare`.
         """
         raise NotImplementedError("Only implemented in child classes")
 
     def _prepare(self):
-        """Set things up before the analysis loop begins"""
+        """
+        Set things up before the analysis loop begins. Note that
+        ``self.results`` is initialized already in :meth:`self.__init__` with an
+        empty instance of :class:`MDAnalysis.analysis.results.Results` object.
+        You can still call your attributes as if they were usual ones,
+        ``Results`` just keeps track of that to be able to run a proper
+        aggregation after a parallel run, if necessary.
```

Comment:
Docstrings first lines have a quasi-meaning or being a one-ish line explanation of what the thing does. We also have a handy `Notes` section in sphinx.

```suggestion
        Set things up before the analysis loop begins. 
        
        Notes
        -----
        ``self.results`` is initialized already in :meth:`self.__init__` with an
        empty instance of :class:`MDAnalysis.analysis.results.Results` object.
        You can still call your attributes as if they were usual ones,
        ``Results`` just keeps track of that to be able to run a proper
        aggregation after a parallel run, if necessary.
```